cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3850 @@
+                std::vector<Datum> result;
+                for(const int &col : column_indices_)
+                    result.push_back(fields[col]);
+                ARROW_ASSIGN_OR_RAISE(ExecBatch batch, ExecBatch::Make(std::move(result)));
+                output_callback(std::move(batch));
+                finished_callback(static_cast<int64_t>(1));
+                return Status::OK();
+            }
+
+            std::shared_ptr<Schema> schema() const override
+            {
+                return schema_;
+            }
+
+            random::pcg32_fast rng_;
+
+            static constexpr size_t kRowCount = 5;
+            static constexpr int32_t kNameByteWidth = 25;
+            const int32_t kRegionKey[kRowCount] = { 0, 1, 2, 3, 4 };
+            const char *kRegionNames[kRowCount] =
+            {
+                "AFRICA", "AMERICA", "ASIA", "EUROPE", "MIDDLE EAST"
+            };
+
+            struct REGION
+            {
+                enum
+                {
+                    R_REGIONKEY,
+                    R_NAME,
+                    R_COMMENT,
+                    kNumColumns,
+                };
+            };
+
+            const std::unordered_map<std::string, int> kNameMap =
+            {
+                { "R_REGIONKEY", REGION::R_REGIONKEY },
+                { "R_NAME", REGION::R_NAME },
+                { "R_COMMENT", REGION::R_COMMENT },
+            };
+
+            const std::vector<std::shared_ptr<DataType>> kTypes =
+            {
+                int32(),
+                fixed_size_binary(kNameByteWidth),
+                utf8(),
+            };
+
+            std::shared_ptr<Schema> schema_;
+            std::vector<int> column_indices_;
+        };
+
+        class TpchNode : public ExecNode
+        {
+        public:
+            TpchNode(ExecPlan *plan,
+                     const char *name,
+                     std::unique_ptr<TpchTableGenerator> generator)
+                : ExecNode(plan, {}, {}, generator->schema(), /*num_outputs=*/1),
+                  name_(name),
+                  generator_(std::move(generator))
+            {
+            }
+
+            const char *kind_name() const override
+            {
+                return name_;
+            }
+
+            [[noreturn]]
+            static void NoInputs()
+            {
+                Unreachable("TPC-H node should never have any inputs");
+            }
+
+            [[noreturn]]
+            void InputReceived(ExecNode *, ExecBatch) override
+            {
+                NoInputs();
+            }
+
+            [[noreturn]]
+            void ErrorReceived(ExecNode *, Status) override
+            {
+                NoInputs();
+            }
+
+            [[noreturn]]
+            void InputFinished(ExecNode *, int) override
+            {
+                NoInputs();
+            }
+
+            Status StartProducing() override
+            {
+                finished_ = Future<>::Make();
```

Comment:
Minor nit: There's no need to do this.  Arguably, the base class shouldn't contain `finished_` while leaving `finished()` pure virtual.